### DISCUSSION TITLE
.redraw() not affected by .activate()/.deactivate()?

### DISCUSSION THREAD

**Discussion by 4JX** [fltk-rs/fltk-rs#969] [Q&A] [answered]
Having a little trouble here, calling redraw after activating/deactivating something does not seem to update its "opacity"/the gui until I move the mouse.

Relevant code:
https://github.com/4JX/L5P-Keyboard-RGB/blob/7374764a5d6dc6a7ac7f53838a3a1eabfceb0cf0/src/gui/keyboard_manager.rs#L67
https://github.com/4JX/L5P-Keyboard-RGB/blob/redraw-test/src/gui/keyboard_color_tiles.rs

**Comment by MoAlyousef** [comment, marked as answer]
Hi
Looking at the code, since you send messages in the handle() method (which is activated on events), you would need to basically call `app::awake()` after `tx.send(message).unwrap();`
Normally you shouldn't need to call .redraw(). 
The advantage with FLTK's own channel is that it internally calls awake. With other channels, you would need to call awake in the case of  sending messages in the handle() method.

**Reply by 4JX** [reply]
When using FLTK's channel with `app::App::wait()` should the messages be sent with `.emit()` from `WidgetExt` or is it more of a convenience method? What's the difference between `app::wait()` and `app::App::wait()`?

**Reply by MoAlyousef** [reply]
You can still use Sender::send(). emit() just uses Sender::send from the callback(). 
There’s no difference between the 2 wait() functions. You also don’t necessarily need to receive messages in the app wait(), you can do so in a separate thread or in the app::add_idle() method.

**Reply by 4JX** [reply]
Changed it to use the fltk channel, still does not update on send (only after mouse movement), might be missing something obvious 😅.
![gif](https://user-images.githubusercontent.com/79868816/139707705-efe3d308-ee07-432a-b019-3eebe97f48ed.gif)

https://github.com/4JX/L5P-Keyboard-RGB/commit/ec8302bf2c85452d1eb4d0c0e805e53489fcdfcc

**Reply by MoAlyousef** [reply]
Hmm what if you add app::awake(); after each send, does it change anything?

Normally you shouldn’t need to.

**Reply by 4JX** [reply]
Adding that seems to prevent messages from being sent (or at least successfully received in the loop) at all, the GUI does not update regardless of mouse movement.

**Reply by MoAlyousef** [reply]
Try adding the app::awake() after the reception of the messages:
```rust
use fltk::{prelude::*, *};

#[derive(Clone)]
enum Message {
    Clicked,
    Done,
}

fn main() {
    let app = app::App::default();
    let (tx, rx) = app::channel();
    let mut wind = window::Window::default().with_size(400, 300);
    let mut frame = frame::Frame::default().with_size(200, 100).center_of(&wind).with_label("0");
    let mut but = button::Button::new(150, 50, 100, 40, "Increment!");
    let mut but2 = button::Button::new(150, 210, 100, 40, "Break");

    wind.end();
    wind.show();

    but.handle({
        let tx = tx.clone();
        move |_, ev| match ev {
        enums::Event::Released => {
            tx.send(Message::Clicked);
            true
        }
        _ => false,
    }});

    but2.handle(move |_, ev| match ev {
        enums::Event::Released => {
            tx.send(Message::Done);
            true
        }
        _ => false,
    });

    std::thread::spawn(move || {
        loop {
            if let Some(msg) = rx.recv() {
                match msg {
                    Message::Clicked => frame.set_label(&(frame.label().parse::<i32>().unwrap() + 1).to_string()),
                    Message::Done => break,
                }
                app::awake();
            }
            app::sleep(0.03);
        }
    });
    app.run().unwrap();
}
```

**Reply by 4JX** [reply]
It now works- mostly. The GUI updates fine for the basic effects, however when a while loop is involved it fails to send (or, again, receive) the message until clicked again. See gif:
![Peek 2021-11-01 20-56](https://user-images.githubusercontent.com/79868816/139733676-a5bf3ed2-8830-43c0-9c07-83879717825f.gif)
- "Callback x" is from `.set_callback()`
- "UpdateEffect message x" is from the loop receiving the messages
- "Lightning loop" / "Escaped lightning effect loop" when looping or after exiting the while loop.

When I select `Static` after `Lightning` it does indeed execute the callback and set the `AtomicBool` (loop ends) but theres no message received until I click it again.

This is with one single `awake()` inside the `if let` as the example shows.

**Reply by MoAlyousef** [reply]
It doesn't seem that the stop signal is changed within the Effects::Lightning loop, that's why it keeps looping. Maybe the stop_signal should be updated in the transition_colors_to() method?

**Reply by 4JX** [reply]
Disregard that last one, messed up my code logic while playing around with updating the gui/awake :p

GUI part of the problem should be solved. Will still need to manually awake in the case I'm entering a while loop since those wont hit the one at the bottom of the main loop until they've exited (the one that handles the messages).
